File: novelos-longform/novelos-app/src-tauri/src/db/repos/chapter.rs
use super::{db_err, Repository};
use rusqlite::Connection;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ChapterRow {
    pub id: String,
    pub chapter_number: i64,
    pub title: Option<String>,
    pub status: String,
    pub draft_text: Option<String>,
    pub final_text: Option<String>,
    pub word_count: Option<i64>,
    pub task_id: Option<String>,
    pub compiler_status: Option<String>,
    pub review_status: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

const SELECT_COLS: &str = "id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at";

fn map_row(row: &rusqlite::Row) -> rusqlite::Result<ChapterRow> {
    Ok(ChapterRow {
        id: row.get(0)?,
        chapter_number: row.get(1)?,
        title: row.get(2)?,
        status: row.get(3)?,
        draft_text: row.get(4)?,
        final_text: row.get(5)?,
        word_count: row.get(6)?,
        task_id: row.get(7)?,
        compiler_status: row.get(8)?,
        review_status: row.get(9)?,
        created_at: row.get(10)?,
        updated_at: row.get(11)?,
    })
}

fn query_list(
    conn: &Connection,
    sql: &str,
    params: impl rusqlite::Params,
) -> Result<Vec<ChapterRow>, String> {
    let mut stmt = conn.prepare(sql).map_err(db_err)?;
    let rows = stmt.query_map(params, map_row).map_err(db_err)?;
    rows.collect::<Result<Vec<_>, _>>().map_err(db_err)
}

pub struct ChapterRepo;

impl Repository<ChapterRow> for ChapterRepo {
    fn find_by_id(conn: &Connection, id: &str) -> Result<Option<ChapterRow>, String> {
        match conn.query_row(
            &format!("SELECT {SELECT_COLS} FROM chapters WHERE id = ?1"),
            [id],
            map_row,
        ) {
            Ok(row) => Ok(Some(row)),
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(db_err(e)),
        }
    }

    fn list(conn: &Connection) -> Result<Vec<ChapterRow>, String> {
        query_list(
            conn,
            &format!("SELECT {SELECT_COLS} FROM chapters ORDER BY chapter_number"),
            [],
        )
    }

    fn delete(conn: &Connection, id: &str) -> Result<(), String> {
        conn.execute("DELETE FROM chapter_versions WHERE chapter_id = ?1", [id])
            .map_err(db_err)?;
        conn.execute("DELETE FROM chapters WHERE id = ?1", [id])
            .map_err(db_err)?;
        Ok(())
    }
}

impl ChapterRepo {
    pub fn find_by_number(
        conn: &Connection,
        chapter_number: i64,
    ) -> Result<Option<ChapterRow>, String> {
        match conn.query_row(
            &format!("SELECT {SELECT_COLS} FROM chapters WHERE chapter_number = ?1"),
            [chapter_number],
            map_row,
        ) {
            Ok(row) => Ok(Some(row)),
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(db_err(e)),
        }
    }

    pub fn update_draft(
        conn: &Connection,
        chapter_number: i64,
        draft_text: &str,
        now: &str,
    ) -> Result<(), String> {
        let word_count = draft_text.chars().count() as i64;
        conn.execute(
            "UPDATE chapters SET draft_text = ?1, word_count = ?2, status = 'drafting', updated_at = ?3 WHERE chapter_number = ?4",
            rusqlite::params![draft_text, word_count, now, chapter_number],
        )
        .map_err(db_err)?;
        Ok(())
    }

    pub fn list_by_status(conn: &Connection, status: &str) -> Result<Vec<ChapterRow>, String> {
        query_list(
            conn,
            &format!(
                "SELECT {SELECT_COLS} FROM chapters WHERE status = ?1 ORDER BY chapter_number"
            ),
            [status],
        )
    }

    pub fn get_chapter_id(
        conn: &Connection,
        chapter_number: i64,
    ) -> Result<Option<String>, String> {
        match conn.query_row(
            "SELECT id FROM chapters WHERE chapter_number = ?1",
            [chapter_number],
            |r| r.get(0),
        ) {
            Ok(id) => Ok(Some(id)),
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(db_err(e)),
        }
    }
}

// ─── Chapter Versions ───

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ChapterVersionRow {
    pub id: String,
    pub chapter_id: String,
    pub version_no: i64,
    pub content_type: String,
    pub content: String,
    pub diff_summary: Option<String>,
    pub created_by: Option<String>,
    pub created_at: String,
}

fn map_version_row(row: &rusqlite::Row) -> rusqlite::Result<ChapterVersionRow> {
    Ok(ChapterVersionRow {
        id: row.get(0)?,
        chapter_id: row.get(1)?,
        version_no: row.get(2)?,
        content_type: row.get(3)?,
        content: row.get(4)?,
        diff_summary: row.get(5)?,
        created_by: row.get(6)?,
        created_at: row.get(7)?,
    })
}

pub struct ChapterVersionRepo;

impl ChapterVersionRepo {
    pub fn list_by_chapter(
        conn: &Connection,
        chapter_id: &str,
    ) -> Result<Vec<ChapterVersionRow>, String> {
        let mut stmt = conn
            .prepare("SELECT id, chapter_id, version_no, content_type, content, diff_summary, created_by, created_at FROM chapter_versions WHERE chapter_id = ?1 ORDER BY version_no DESC")
            .map_err(db_err)?;
        let rows = stmt
            .query_map([chapter_id], map_version_row)
            .map_err(db_err)?;
        rows.collect::<Result<Vec<_>, _>>().map_err(db_err)
    }
}

// ─── Chapter Tasks ───

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ChapterTaskRow {
    pub id: String,
    pub chapter_number: i64,
    pub volume_id: Option<String>,
    pub arc_id: Option<String>,
    pub objective: String,
    pub must_progress: Option<String>,
    pub must_recall: Option<String>,
    pub must_avoid: Option<String>,
    pub required_hooks: Option<String>,
    pub required_context: Option<String>,
    pub ending_hook: Option<String>,
    pub status: Option<String>,
    pub created_at: String,
}

fn map_task_row(row: &rusqlite::Row) -> rusqlite::Result<ChapterTaskRow> {
    Ok(ChapterTaskRow {
        id: row.get(0)?,
        chapter_number: row.get(1)?,
        volume_id: row.get(2)?,
        arc_id: row.get(3)?,
        objective: row.get(4)?,
        must_progress: row.get(5)?,
        must_recall: row.get(6)?,
        must_avoid: row.get(7)?,
        required_hooks: row.get(8)?,
        required_context: row.get(9)?,
        ending_hook: row.get(10)?,
        status: row.get(11)?,
        created_at: row.get(12)?,
    })
}

pub struct ChapterTaskRepo;

impl ChapterTaskRepo {
    pub fn list(conn: &Connection, volume_id: Option<&str>) -> Result<Vec<ChapterTaskRow>, String> {
        let sql = if volume_id.is_some() {
            "SELECT id, chapter_number, volume_id, arc_id, objective, must_progress, must_recall, must_avoid, required_hooks, required_context, ending_hook, status, created_at FROM chapter_tasks WHERE volume_id = ?1 ORDER BY chapter_number"
        } else {
            "SELECT id, chapter_number, volume_id, arc_id, objective, must_progress, must_recall, must_avoid, required_hooks, required_context, ending_hook, status, created_at FROM chapter_tasks ORDER BY chapter_number"
        };
        let mut stmt = conn.prepare(sql).map_err(db_err)?;

        let rows = if let Some(vid) = volume_id {
            stmt.query_map([vid], map_task_row).map_err(db_err)?
        } else {
            stmt.query_map([], map_task_row).map_err(db_err)?
        };
        rows.collect::<Result<Vec<_>, _>>().map_err(db_err)
    }
}

File: novelos-longform/novelos-app/src/lib/web-api/chapter.ts
import { webDb } from "../web-db";
import { uuid, now, boolToInt, intToBool, nullIfUndefined, requireProjectId, WebNotSupportedError } from "./index";

import type {
  ChapterTaskInfo,
  ChapterInfo,
  ChapterVersionInfo,
  ChapterSearchResult,
  RecalledContext,
  CharacterInfo,
} from "../tauri";

const CHAPTER_TRANSITIONS: Record<string, string[]> = {
  task_ready: ["drafting"],
  drafting: ["review", "task_ready"],
  review: ["finalized", "drafting", "task_ready"],
  finalized: ["approved", "review"],
  approved: [],
};

export const chapterApi = {
  async listTasks(volumeId?: string): Promise<ChapterTaskInfo[]> {
    return webDb.all<ChapterTaskInfo>(
      `SELECT id, chapter_number, volume_id, arc_id, objective, must_progress, must_recall, must_avoid, required_hooks, required_context, ending_hook, status, created_at
       FROM chapter_tasks WHERE (? IS NULL OR volume_id = ?) ORDER BY chapter_number`,
      [volumeId ?? null, volumeId ?? null],
    );
  },

  async createTask(
    chapterNumber: number, objective: string, volumeId?: string, arcId?: string,
    mustProgress?: string, mustRecall?: string, mustAvoid?: string,
  ): Promise<ChapterTaskInfo> {
    const id = uuid();
    const ts = now();
    const projectId = requireProjectId();
    webDb.run(
      `INSERT INTO chapter_tasks (id, project_id, chapter_number, volume_id, arc_id, objective, must_progress, must_recall, must_avoid, status, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 'pending', ?)`,
      [id, projectId, chapterNumber, nullIfUndefined(volumeId), nullIfUndefined(arcId),
       objective, nullIfUndefined(mustProgress), nullIfUndefined(mustRecall), nullIfUndefined(mustAvoid), ts],
    );
    return webDb.get<ChapterTaskInfo>(
      "SELECT id, chapter_number, volume_id, arc_id, objective, must_progress, must_recall, must_avoid, required_hooks, required_context, ending_hook, status, created_at FROM chapter_tasks WHERE id = ?",
      [id],
    )!;
  },

  async listChapters(): Promise<ChapterInfo[]> {
    return webDb.all<ChapterInfo>(
      "SELECT id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at FROM chapters ORDER BY chapter_number",
    );
  },

  async getChapter(chapterNumber: number): Promise<ChapterInfo> {
    return webDb.get<ChapterInfo>(
      "SELECT id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at FROM chapters WHERE chapter_number = ?",
      [chapterNumber],
    )!;
  },

  async createChapter(chapterNumber: number, title?: string, taskId?: string): Promise<ChapterInfo> {
    const id = uuid();
    const ts = now();
    const projectId = requireProjectId();
    webDb.run(
      "INSERT INTO chapters (id, project_id, chapter_number, title, status, task_id, created_at, updated_at) VALUES (?, ?, ?, ?, 'task_ready', ?, ?, ?)",
      [id, projectId, chapterNumber, nullIfUndefined(title), nullIfUndefined(taskId), ts, ts],
    );
    return webDb.get<ChapterInfo>(
      "SELECT id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at FROM chapters WHERE id = ?",
      [id],
    )!;
  },

  async updateDraft(chapterNumber: number, draftText: string, skipVersion?: boolean): Promise<ChapterInfo> {
    const ts = now();
    const wordCount = draftText.length;
    webDb.run(
      "UPDATE chapters SET draft_text = ?, word_count = ?, status = 'drafting', updated_at = ? WHERE chapter_number = ?",
      [draftText, wordCount, ts, chapterNumber],
    );
    // Insert chapter version only for manual saves
    if (!skipVersion) {
      const chapter = webDb.get<{ id: string }>("SELECT id FROM chapters WHERE chapter_number = ?", [chapterNumber]);
      if (chapter) {
        const nextVer = webDb.get<{ v: number }>(
          "SELECT COALESCE(MAX(version_no), 0) + 1 as v FROM chapter_versions WHERE chapter_id = ?",
          [chapter.id],
        );
        webDb.run(
          "INSERT INTO chapter_versions (id, chapter_id, version_no, content_type, content, created_by, created_at) VALUES (?, ?, ?, 'draft', ?, 'user', ?)",
          [uuid(), chapter.id, nextVer?.v ?? 1, draftText, ts],
        );
      }
    }
    return webDb.get<ChapterInfo>(
      "SELECT id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at FROM chapters WHERE chapter_number = ?",
      [chapterNumber],
    )!;
  },

  async finalize(chapterNumber: number): Promise<ChapterInfo> {
    const ts = now();
    const chapter = webDb.get<ChapterInfo>(
      "SELECT id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at FROM chapters WHERE chapter_number = ?",
      [chapterNumber],
    );
    if (!chapter) throw new Error(`Chapter ${chapterNumber} not found`);
    const finalText = chapter.draft_text ?? "";
    const wordCount = finalText.length;
    webDb.run(
      "UPDATE chapters SET final_text = ?, word_count = ?, status = 'finalized', updated_at = ? WHERE chapter_number = ?",
      [finalText, wordCount, ts, chapterNumber],
    );
    // Insert final version
    const nextVer = webDb.get<{ v: number }>(
      "SELECT COALESCE(MAX(version_no), 0) + 1 as v FROM chapter_versions WHERE chapter_id = ?",
      [chapter.id],
    );
    webDb.run(
      "INSERT INTO chapter_versions (id, chapter_id, version_no, content_type, content, created_by, created_at) VALUES (?, ?, ?, 'final', ?, 'user', ?)",
      [uuid(), chapter.id, nextVer?.v ?? 1, finalText, ts],
    );
    return webDb.get<ChapterInfo>(
      "SELECT id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at FROM chapters WHERE chapter_number = ?",
      [chapterNumber],
    )!;
  },

  async listVersions(chapterNumber: number): Promise<ChapterVersionInfo[]> {
    const chapter = webDb.get<{ id: string }>("SELECT id FROM chapters WHERE chapter_number = ?", [chapterNumber]);
    if (!chapter) return [];
    return webDb.all<ChapterVersionInfo>(
      "SELECT id, chapter_id, version_no, content_type, content, diff_summary, created_by, created_at FROM chapter_versions WHERE chapter_id = ? ORDER BY version_no DESC",
      [chapter.id],
    );
  },

  async rollback(chapterNumber: number, versionNo: number): Promise<ChapterInfo> {
    const chapter = webDb.get<{ id: string }>("SELECT id FROM chapters WHERE chapter_number = ?", [chapterNumber]);
    if (!chapter) throw new Error(`Chapter ${chapterNumber} not found`);
    const version = webDb.get<{ content: string }>(
      "SELECT content FROM chapter_versions WHERE chapter_id = ? AND version_no = ?",
      [chapter.id, versionNo],
    );
    if (!version) throw new Error(`Version ${versionNo} not found`);
    const wordCount = version.content.length;
    webDb.run(
      "UPDATE chapters SET draft_text = ?, word_count = ?, updated_at = ? WHERE chapter_number = ?",
      [version.content, wordCount, now(), chapterNumber],
    );
    return webDb.get<ChapterInfo>(
      "SELECT id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at FROM chapters WHERE chapter_number = ?",
      [chapterNumber],
    )!;
  },

  async searchChapters(query: string): Promise<ChapterSearchResult[]> {
    const term = `%${query}%`;
    const rows = webDb.all<{
      chapter_number: number; title: string | null;
      content: string; status: string; word_count: number | null;
    }>(
      `SELECT chapter_number, title, COALESCE(final_text, draft_text, '') as content, status, word_count
       FROM chapters WHERE final_text LIKE ? OR draft_text LIKE ? OR title LIKE ? ORDER BY chapter_number LIMIT 30`,
      [term, term, term],
    );
    return rows.map((r) => ({
      chapter_number: r.chapter_number,
      title: r.title,
      snippet: r.content.substring(0, 200),
      status: r.status,
      word_count: r.word_count,
    }));
  },

  async recallContext(chapterNumber: number): Promise<RecalledContext> {
    const hardRules = webDb.all<{ rule_name: string; content: string }>(
      "SELECT rule_name, content FROM canon_rules WHERE status = 'active' AND is_hard = 1 ORDER BY rule_name",
    );
    const softRules = webDb.all<{ rule_name: string; content: string }>(
      "SELECT rule_name, content FROM canon_rules WHERE status = 'active' AND is_hard = 0 ORDER BY rule_name LIMIT 10",
    );
    const charStates = webDb.all<{ name: string; level_state: string | null; emotion_state: string | null; goal_state: string | null }>(
      "SELECT c.name, cs.level_state, cs.emotion_state, cs.goal_state FROM characters c LEFT JOIN character_states cs ON cs.character_id = c.id WHERE c.status = 'active' LIMIT 15",
    );
    const foreshadows = webDb.all<{ title: string; seed_chapter: number }>(
      "SELECT title, seed_chapter FROM foreshadow_items WHERE status = 'planted' ORDER BY seed_chapter LIMIT 10",
    );

    return {
      hard_rules: hardRules.map((r) => `${r.rule_name}: ${r.content}`),
      soft_rules: softRules.map((r) => `${r.rule_name}: ${r.content}`),
      character_states: charStates.map((c) => `${c.name}: ${[c.level_state, c.emotion_state, c.goal_state].filter(Boolean).join(", ")}`),
      open_foreshadows: foreshadows.map((f) => `${f.title} (seed: ch${f.seed_chapter})`),
      total_tokens_estimate: 0,
    };
  },

  async listCharacters(): Promise<CharacterInfo[]> {
    return webDb.all<CharacterInfo>(
      "SELECT id, name, alias, role_type, identity_core, persona_core, soul_template_id, soul_json, taboo_rules, core_motivation, status, created_at, updated_at FROM characters ORDER BY name",
    );
  },

  async createCharacter(name: string, roleType?: string, soulJson?: string): Promise<CharacterInfo> {
    const id = uuid();
    const ts = now();
    const projectId = requireProjectId();
    const role = roleType ?? "supporting";
    const soul = soulJson ?? "{}";
    webDb.run(
      "INSERT INTO characters (id, project_id, name, role_type, soul_json, status, created_at, updated_at) VALUES (?, ?, ?, ?, ?, 'active', ?, ?)",
      [id, projectId, name, role, soul, ts, ts],
    );
    return webDb.get<CharacterInfo>(
      "SELECT id, name, alias, role_type, identity_core, persona_core, soul_template_id, soul_json, taboo_rules, core_motivation, status, created_at, updated_at FROM characters WHERE id = ?",
      [id],
    )!;
  },

  async updateCharacter(
    id: string, name?: string, soulJson?: string, roleType?: string,
    identityCore?: string, personaCore?: string, coreMotivation?: string, tabooRules?: string,
  ): Promise<void> {
    const ts = now();
    if (name !== undefined) webDb.run("UPDATE characters SET name = ?, updated_at = ? WHERE id = ?", [name, ts, id]);
    if (soulJson !== undefined) webDb.run("UPDATE characters SET soul_json = ?, updated_at = ? WHERE id = ?", [soulJson, ts, id]);
    if (roleType !== undefined) webDb.run("UPDATE characters SET role_type = ?, updated_at = ? WHERE id = ?", [roleType, ts, id]);
    if (identityCore !== undefined) webDb.run("UPDATE characters SET identity_core = ?, updated_at = ? WHERE id = ?", [identityCore, ts, id]);
    if (personaCore !== undefined) webDb.run("UPDATE characters SET persona_core = ?, updated_at = ? WHERE id = ?", [personaCore, ts, id]);
    if (coreMotivation !== undefined) webDb.run("UPDATE characters SET core_motivation = ?, updated_at = ? WHERE id = ?", [coreMotivation, ts, id]);
    if (tabooRules !== undefined) webDb.run("UPDATE characters SET taboo_rules = ?, updated_at = ? WHERE id = ?", [tabooRules, ts, id]);
  },

  async deleteCharacter(id: string): Promise<void> {
    webDb.run("DELETE FROM characters WHERE id = ?", [id]);
  },

  async transitionState(chapterNumber: number, newStatus: string): Promise<ChapterInfo> {
    webDb.run(
      "UPDATE chapters SET status = ?, updated_at = ? WHERE chapter_number = ?",
      [newStatus, now(), chapterNumber],
    );
    return webDb.get<ChapterInfo>(
      "SELECT id, chapter_number, title, status, draft_text, final_text, word_count, task_id, compiler_status, review_status, created_at, updated_at FROM chapters WHERE chapter_number = ?",
      [chapterNumber],
    )!;
  },

  async getValidTransitions(chapterNumber: number): Promise<string[]> {
    const chapter = webDb.get<{ status: string }>(
      "SELECT status FROM chapters WHERE chapter_number = ?", [chapterNumber],
    );
    if (!chapter) return [];
    return CHAPTER_TRANSITIONS[chapter.status] ?? [];
  },

  async setCompileStatus(chapterNumber: number, compilerStatus: string): Promise<void> {
    webDb.run(
      "UPDATE chapters SET compiler_status = ?, updated_at = ? WHERE chapter_number = ?",
      [compilerStatus, now(), chapterNumber],
    );
  },

  async setReviewStatus(chapterNumber: number, reviewStatus: string): Promise<void> {
    webDb.run(
      "UPDATE chapters SET review_status = ?, updated_at = ? WHERE chapter_number = ?",
      [reviewStatus, now(), chapterNumber],
    );
  },

  async saveTitle(chapterNumber: number, title: string): Promise<void> {
    webDb.run(
      "UPDATE chapters SET title = ?, updated_at = ? WHERE chapter_number = ?",
      [title, now(), chapterNumber],
    );
  },
};
